Typings: Add Observer interface and improve typings for Auth, Storage, and Messaging

jsayol opened the pull request:
### Discussion
Fixes issue #54 

### Testing
n/a

### API Changes
n/a

jsayol commented:
Out of curiosity: these typing files seem to be auto-generated. Was that a one-time thing, or is there a process to regenerate them when there's changes to the API?

bojeil-google commented on typings/app.d.ts:
```diff
@@ -123,7 +123,7 @@ declare namespace firebase.auth {
     getRedirectResult ( ) : firebase.Promise < any > ;
-    onAuthStateChanged (nextOrObserver : Object , error ? : (a : firebase.auth.Error ) => any , completed ? : ( ) => any ) : ( ) => any ;
+    onAuthStateChanged (nextOrObserver : Object | ((user: firebase.User | null) => any), error ? : (a : firebase.auth.Error ) => any , completed ? : ( ) => any ) : ( ) => any ;
     onIdTokenChanged (nextOrObserver : Object , error ? : (a : firebase.auth.Error ) => any , completed ? : ( ) => any ) : ( ) => any ;
```
While you're at it, can you make the same change here too?

jsayol replied:
Good point. I'll check if there's any other places too.

How about also introducing an `Observer` interface to use instead of the more generic `Object`?
```ts
interface Observer {
  next?: (user: firebase.User | null) => any;
  error?: (error: firebase.auth.Error) => any;
  complete?: () => any;
}
```

bojeil-google commented on typings/firebase.d.ts:
```diff
@@ -123,7 +123,7 @@ declare namespace firebase.auth {
     getRedirectResult ( ) : firebase.Promise < any > ;
-    onAuthStateChanged (nextOrObserver : Object , error ? : (a : firebase.auth.Error ) => any , completed ? : ( ) => any ) : ( ) => any ;
+    onAuthStateChanged (nextOrObserver : Object | ((user: firebase.User | null) => any), error ? : (a : firebase.auth.Error ) => any , completed ? : ( ) => any ) : ( ) => any ;
     onIdTokenChanged (nextOrObserver : Object , error ? : (a : firebase.auth.Error ) => any , completed ? : ( ) => any ) : ( ) => any ;
```
same here.

jsayol replied:
Both done.

jshcrowthe commented:
@gauntface Can you also take a look at this?

sphippen commented on typings/app.d.ts:
```diff
@@ -76,6 +76,12 @@ declare namespace firebase {
   }
 
+  interface Observer<V, E> {
+    next?: (value: V | null) => any;
```
In a few places inside the SDK, these are typed with a void return value (https://github.com/firebase/firebase-js-sdk/blob/657055d3c81101d55e86baac0a49d7cfd0a25863/src/app/subscribe.ts#L16, https://github.com/firebase/firebase-js-sdk/blob/657055d3c81101d55e86baac0a49d7cfd0a25863/src/storage/implementation/observer.ts#L18). Let's make the return types here "void" instead of "any" for consistency.

sphippen replied:
(note: also applies to the return types in the function signatures below)

jsayol replied:
Done.

jsayol commented:
Should these same changes be reflected in the type annotation comments of the _externs_ files?

sphippen commented:
Hmm, forgot about those. As far as I can tell we've been keeping the TS<->Closure typings in sync, so the Closure externs files should also be updated.

jsayol commented:
@sphippen I just pushed the changes to the _externs_ to get them in sync with the typings. I'm not extremely familiar with Closure compiler annotations so make sure to review those carefully, please :)

I've also made a few more improvements to the typings while I was at it. I added the `NextFn`, `ErrorFn`, and `CompleteFn` types (they were already present in the _externs_) and I modified the signature of the `firebase.storage.UploadTask#on()` method to reflect the two different usages.

sphippen commented on externs/firebase-app-externs.js:
```diff
@@ -273,3 +273,38 @@ firebase.Promise.reject = function(error) {};
+
+firebase.Observer.prototype.complete = function() {};
+
+/**
```
Unfortunately Closure doesn't support templated typedefs:

```
../../externs/firebase-app-externs.js:296: WARNING - Bad type annotation. type annotation incompatible with other annotations. See https://github.com/google/closure-compiler/wiki/Bad-Type-Annotation for more information.
 * @typedef {function(?V): void}
   ^

../../externs/firebase-app-externs.js:298: WARNING - Misplaced template annotation. @template is only allowed in class, constructor, interface, function or method declarations
firebase.NextFn;
^^^^^^^^^^^^^^^

../../externs/firebase-app-externs.js:302: WARNING - Bad type annotation. type annotation incompatible with other annotations. See https://github.com/google/closure-compiler/wiki/Bad-Type-Annotation for more information.
 * @typedef {function(!E): void}
   ^

../../externs/firebase-app-externs.js:304: WARNING - Misplaced template annotation. @template is only allowed in class, constructor, interface, function or method declarations
firebase.ErrorFn;
^^^^^^^^^^^^^^^^
```

It's probably best to leave the NextFn and ErrorFn arguments as raw "!function(...)" annotations instead of adding the typedefs here. I'd be fine if you left CompleteFn and Unsubscribe in as typedefs and added a comment about the templated typedefs being unfeasible or you just removed all of them. Either one is good with me.

jsayol replied:
Ah well, too bad. I removed the templates from NextFn and ErrorFn and moved the typedefs back to the INTERNAL file, where they're still used. I replaced all the other uses with the original `!function(...)` annotations.

sphippen commented on externs/firebase-app-externs.js:
```diff
@@ -275,3 +275,26 @@ firebase.Promise.reject = function(error) {};
+
+/**
+ * @template V
+ * @param {?V} value
```
Just one last change, put the "@template V" and "@template E" annotations in the same comment block as the "@interface" instead of on the individual methods.

jsayol replied:
Like this?

```js
/**
 * @template V, E
 * @interface
 **/
firebase.Observer = function() {};

/**
 * @param {?V} value
 */
firebase.Observer.prototype.next = function(value) {};

/**
 * @param {!E} error
 */
firebase.Observer.prototype.error = function(error) {};

firebase.Observer.prototype.complete = function() {};
```

jsayol replied:
Done

sphippen commented:
Awesome, looks good to me.
Also @jsayol, re: automatic generation of the typings, they have always been maintained by hand. If I remember correctly the TypeScript typings were based on a machine-converted form of the Closure externs files, but we had to do some manual cleanup afterward and have been maintaining them by hand since.

Thanks for the PR!
